### PR TITLE
feat(login): verify logins exist

### PR DESCRIPTION
**Pull request by slin5299**

### CODE REVIEW

File: src/Exceptions/MissingLoginException.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  using System;
+
+
+  /// <summary>
+  ///   A base exception for the AgDatabaseMove library
+  /// </summary>
+  public class MissingLoginException : AgDatabaseMoveException
+  {
+    public MissingLoginException(string message) : base(message) { }
+  }
```
Comment (AdityaNPL) [inline src/Exceptions/MissingLoginException.cs:12]: The summary might need to be changed

Comment (AdityaNPL) [reply src/Exceptions/MissingLoginException.cs:12]: Same for the other exceptions too

File: src/SmoFacade/Listener.cs [hunk 2]
```diff
@@ -89,6 +98,8 @@ public Listener(SqlConnectionStringBuilder connectionStringBuilder, string crede
 
     public AvailabilityGroup AvailabilityGroup { get; set; }
 
+    public IEnumerable<Server> AllInstances => _allInstances;
+
```
Comment (AdityaNPL) [inline src/SmoFacade/Listener.cs:102]: You might be able to do a join b/w primaries and secondaries - not 100% sure tho??

**Review by AdityaNPL — COMMENTED** [review]
Looks good!
I just gave it a quick glance over, I'll leave it to Alex/Ryan for the final reviews.

Some quick suggestions from my side:

File: src/Exceptions/MissingLoginException.cs (hunk 1 above)
Comment (slin5299) [reply src/Exceptions/MissingLoginException.cs]: yup, done.

File: src/SmoFacade/Listener.cs (hunk 2 above)
Comment (slin5299) [reply src/SmoFacade/Listener.cs]: I removed this since it is no longer needed.

Comment (alexirion10) [reply src/SmoFacade/Listener.cs]: Yeah I'm not sure we need an additional variable for this.  You should be able to join the primary and secondary members inline as a property

File: src/SmoFacade/Listener.cs [hunk 3]
```diff
@@ -74,10 +74,12 @@ public Listener(SqlConnectionStringBuilder connectionStringBuilder, string crede
...
-      foreach(var secondaryName in secondaryNames)
-        _secondaries.Add(AgInstanceNameToServer(ref connectionStringBuilder,
-                                                secondaryName,
-                                                credentialName));
+      foreach(var secondaryName in secondaryNames) {
+        var secondary = AgInstanceNameToServer(ref connectionStringBuilder,
+                                               secondaryName,
+                                               credentialName);
+        _secondaries.Add(secondary);
+      }
```
Comment (alexirion10) [inline src/SmoFacade/Listener.cs:82]: can you revert this back :)

**Review by alexirion10 — COMMENTED** [review]
Looks pretty good to me!  Have you had a chance to test this on any clusters yet?

File: src/SmoFacade/Listener.cs (hunk 3 above)
Comment (slin5299) [reply src/SmoFacade/Listener.cs]: done

File: src/AgDatabase.cs [hunk 4]
```diff
@@ -233,5 +237,48 @@ public void MultiUserMode()
...
+        }
+      });
+
+      if(exceptions.Count > 0) throw new AggregateException(exceptions);
+    }
+
+    public byte[] GetSid(string dbName)
+    {
+      return _listener.Primary.Logins.Single(d => d.Name == dbName).Sid;
+    }
```
[the remark this answers is not in the document]
Comment (slin5299) [reply src/AgDatabase.cs]: done

File: src/AgDatabase.cs [hunk 5]
```diff
@@ -12,6 +12,8 @@ namespace AgDatabaseMove
   using System.Data.SqlClient;
   using System.Linq;
   using System.Threading;
+  using Exceptions;
+  using Microsoft.SqlServer.Management.SqlParser.SqlCodeDom;
```
[the remark this answers is not in the document]
Comment (slin5299) [reply src/AgDatabase.cs]: I don't remember, it's been removed

**Comment by slin5299** [comment]
It would be hard to test on a cluster since AgDatabase doesn't have access to admin credentials and Catalogue. I've been hard coding individual listeners and databases to run tests.

**Comment by alexirion10** [comment]
@slin5299 - yep exactly, that's good!  Just wanted to make sure it was tested on a real cluster

File: src/AgDatabase.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexirion10) [reply src/AgDatabase.cs]: So I actually realized we already have a function that returns `LoginProperty`... actually a list of all login-properties, but they can still be filtered out in the same way. Right?
So I don't think we need this at all!  my b

[the remark this answers is not in the document]
Comment (slin5299) [reply src/AgDatabase.cs]: Yep, `AssociatedLogins()` returns an IEnumerable, which can be filtered. I'll remove this function.

**Review by alexirion10 — APPROVED** [review]

**Pull request merged**